source/common/network/io_socket_handle_impl.cc

```diff
@@ -236,7 +239,19 @@ maybeGetDstAddressFromHeader(const cmsghdr& cmsg, uint32_t self_port, os_fd_t fd
     ipv6_addr->sin6_family = AF_INET6;
     ipv6_addr->sin6_addr = info->ipi6_addr;
     ipv6_addr->sin6_port = htons(self_port);
-    return Address::addressFromSockAddrOrDie(ss, sizeof(sockaddr_in6), fd, v6only);
+    if (recent_received_addresses_ == nullptr) {
+      return Address::addressFromSockAddrOrDie(ss, sizeof(sockaddr_in6), fd, v6only);
+    }
+    quic::QuicSocketAddress quic_address(ss);
+    auto it = recent_received_addresses_->Lookup(quic_address);
+    if (it == recent_received_addresses_->end()) {
+      Address::InstanceConstSharedPtr new_address =
+          Address::addressFromSockAddrOrDie(ss, sizeof(sockaddr_in6), fd, v6only);
+      recent_received_addresses_->Insert(
+          quic_address, std::make_unique<Address::InstanceConstSharedPtr>(new_address));
+      return new_address;
+    }
+    return *it->second;
```

Comment:
nit: since this is the smaller codepath, let's inver the conditional and reduce the nesting a tad:

```
    if (it != recent_received_addresses_->end()) {
      return *it->second;
    }
    Address::InstanceConstSharedPtr new_address =
        Address::addressFromSockAddrOrDie(ss, sizeof(sockaddr_in6), fd, v6only);
    recent_received_addresses_->Insert(
        quic_address, std::make_unique<Address::InstanceConstSharedPtr>(new_address));
    return new_address;
```